### PR TITLE
Ipynb estimate params

### PR DESCRIPTION
**Pull request by mrzkp**
Jupyter Notebook for estimate_params example.

### CODE REVIEW

File: examples/param_est.ipynb [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+   "source": [
+    "from prog_models.models.thrown_object import ThrownObject"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we can build a model with your best guess in parameters.\n",
```
Comment (kjjarvis) [inline examples/param_est.ipynb:65]:
```suggestion
    "Now we can build a model with a best guess for the parameters.\n",
```

File: examples/param_est.ipynb [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+   "source": [
+    "m = ThrownObject(thrower_height=20)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, we will collect data from the model of the system. Let's pretend we threw the ball once, and collected position measurments."
```
Comment (kjjarvis) [inline examples/param_est.ipynb:84]:
```suggestion
    "Next, we will collect data from the system. Let's pretend we threw the ball once, and collected position measurements."
```
No model at this point, just 'data'

File: examples/param_est.ipynb [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+   "source": [
+    "Sure enough- parameter estimation determined that the thrower's height wasn't 20 m, instead was closer to 1.9m, a much more reasonable height!"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`estimate_params()` creates a structure of 'runs' by constructing each index of times, inputs, and outputs and placing them into a tuple.\n",
```
Comment (kjjarvis) [inline examples/param_est.ipynb:198]:
```suggestion
    "`estimate_params()` creates a structure of 'runs' by taking each index of times, inputs, and outputs and placing them into a tuple.\n",
```

File: examples/param_est.ipynb [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    "\n",
+    "                `runs = [(t, u, z) for t, u, z in zip(times, inputs, outputs)]`\n",
+    "\n",
+    "\n",
+    "Using our optimization function, which runs `calc_error()` as a subroutine (more information about `calc_error()` found in our Calculating Error Example), given each of the runs.\n",
```
Comment (kjjarvis) [inline examples/param_est.ipynb:203]: This sentence doesn't make sense to me. Can we reword it?

File: examples/param_est.ipynb [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    "\n",
+    "You can also adjust the metric that is used to estimate parameters by setting the error_method to a different `calc_error()` method.\n",
+    "e.g., m.estimate_params([(times, inputs, outputs)], keys, dt=0.01, error_method='MAX_E')\n",
+    "Default is Mean Squared Error (MSE)\n",
+    "See calc_error method for list of options."
```
Comment (kjjarvis) [inline examples/param_est.ipynb:208]: This is a really nice additional note. Should we include a line of code here to illustrate it? (basically pull out line 206 and make it an executable statement?)

File: examples/param_est.ipynb [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we can build a model with your best guess in parameters.\n",
+    "\n",
+    "We will use a ThrownObject Model and guess that our thrower is 20 meters tall. However, given our times, inputs and outputs, we can clearly tell this is obviously not true! Let's see if parameter estimation can fix this!"
```
Comment (kjjarvis) [inline examples/param_est.ipynb:67]: This comment confused me a bit because we don't have times, inputs, or outputs at this point. Maybe we could reword it? It is an unreasonable height for a human, I think that is enough

File: examples/param_est.ipynb (hunk 4 above)
Comment (kjjarvis) [inline examples/param_est.ipynb:203]: What is the motivation for including these lines?

**Review by kjjarvis — CHANGES_REQUESTED** [review]
Great job on a first jupyter notebook example!

File: examples/param_est.ipynb [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The goal of this notebook is to instruct ProgPy users on how to use the estimate_params feature for PrognosticModels.\n",
+    "\n",
+    "First some background. Parameter estimation is used to tune a general model to match the behavior of a specific system. For example, parameters of the general battery model can be used to configure the model to describe the behavior of a specific battery.\n",
```
Comment (teubert) [inline examples/param_est.ipynb:18]:
```suggestion
    "First some background. Parameter estimation is used to tune the parameters of a general model so its behavior matches the behavior of a specific system. For example, parameters of the battery model can be tuned to configure the model to describe the behavior of a specific battery.\n",
```

File: examples/param_est.ipynb [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    "\n",
+    "Generally, parameter estimation is done by tuning the parameters of the model so that simulation best matches the behavior observed in some available data. In ProgPy, this is done using the prog_models.PrognosticsModel.estimate_params() method. This method takes input and output data from one or more runs, and uses scipy.optimize.minimize function to estimate the parameters of the model. For more information, refer to our Documentation (https://nasa.github.io/progpy/prog_models_guide.html#parameter-estimation)\n",
```
Comment (teubert) [inline examples/param_est.ipynb:20]:
```suggestion
    "Generally, parameter estimation is done by tuning the parameters of the model so that simulation best matches the behavior observed in some available data. In ProgPy, this is done using the prog_models.PrognosticsModel.estimate_params() method. This method takes input and output data from one or more runs, and uses scipy.optimize.minimize function to estimate the parameters of the model. For more information, refer to our Documentation [here](https://nasa.github.io/progpy/prog_models_guide.html#parameter-estimation)\n",
```

File: examples/param_est.ipynb [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    "\n",
+    "A few definitions:\n",
+    "* __`keys`__ `(list[str])`: Parameter keys to optimize\n",
+    "* __`times`__ `(list[float])`: Array of times for each sample\n",
```
Comment (teubert) [inline examples/param_est.ipynb:24]:
```suggestion
    "* __`times`__ `(list[float])`: Array of times for each run\n",
```

File: examples/param_est.ipynb [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+    "* __`bounds`__ `(tuple or dict, optional)`: Bounds for optimization in format ((lower1, upper1), (lower2, upper2), ...) or {key1: (lower1, upper1), key2: (lower2, upper2), ...}\n",
+    "* __`options`__ `(dict, optional)`: Options passed to optimizer. See scipy.optimize.minimize for options"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Example"
```
Comment (teubert) [inline examples/param_est.ipynb:38]:
```suggestion
    "### Simple Example"
```

File: examples/param_est.ipynb [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we will show an example demonstrating the model parameter estimation feature.\n",
```
Comment (teubert) [inline examples/param_est.ipynb:46]:
```suggestion
    "Now we will show an example demonstrating the model parameter estimation feature. In this example, we will be estimating the parameters for a model from data . In general, the data will usually be collected from the physical system or from a different model (model surrogacy). \n",
```

File: examples/param_est.ipynb [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+    "\n",
+    "First, we will import the necessary modules from the ProgPy Package."
```
Comment (teubert) [inline examples/param_est.ipynb:48]:
```suggestion
    "First, we will import a model from the ProgPy Package. For this example we're using the simple ThrownObject model."
```

File: examples/param_est.ipynb [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we can build a model with a best guess for the parameters.\n",
+    "\n",
+    "We will use a ThrownObject Model and guess that our thrower is 20 meters tall. However, given our times, inputs and outputs, we can clearly tell this is obviously not true! Let's see if parameter estimation can fix this!"
```
Comment (teubert) [inline examples/param_est.ipynb:67]:
```suggestion
    "We will use a guess that our thrower is 20 meters tall. However, given our times, inputs, and outputs, we can clearly tell this is not true! Let's see if parameter estimation can fix this!"
```

File: examples/param_est.ipynb [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+    "    print(\"-\", key, m.parameters[key])\n",
+    "print(' Error: ', m.calc_error(times, inputs, outputs, dt=1e-4))"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, we will run `estimate_params()` with the data."
```
Comment (teubert) [inline examples/param_est.ipynb:153]:
```suggestion
    "Notice that the error is quite high. This indicates that the parameters are not accurate",
    "",
    "Now, we will run `estimate_params()` with the data to correct these parameters."
```

File: examples/param_est.ipynb (hunk 4 above)
Comment (teubert) [reply examples/param_est.ipynb:203]: As discussed in our meeting, let's create a new example with multiple runs to introduce the concept of runs. I have ideas for this, so let me know if you need help. 

Then we can reword this to just describe how calc_error is applied to each run independently, without the definition of runs.

**Review by teubert — CHANGES_REQUESTED** [review]
This is really great! A great addition. 

Just a few comments/suggestions.

File: examples/param_est.ipynb [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+    "* Why multiple runs with the noise, both noises work, run estimate_params(), not very good of a job with one run.\n",
+    "* Add tolerance, and change the error calculation metric, (just a note for error calculation metric)."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": []
+  }
```
Comment (teubert) [inline examples/param_est.ipynb:419]:
```suggestion
```

**Comment by teubert** [comment]
Also, let's get rid of the estimate_params.py example. It's replaced by this

**Review by teubert — APPROVED** [review]

File: tests/test_estimate_params.py [hunk 16]
```diff
@@ -944,8 +944,8 @@ def test_tolerance(self):
 
         # When tolerance is empty list
         m.parameters['thrower_height'] = 3.1
-        m.parameters['throwing_speed'] = 29
-        m.parameters['g'] = 10
+        # m.parameters['throwing_speed'] = 29
+        # m.parameters['g'] = 10
```
Comment (kjjarvis) [inline tests/test_estimate_params.py:948]: What was the reason for commenting these out? Should we just delete them?

**Review by kjjarvis — CHANGES_REQUESTED** [review]
Most of these are just small changes, but I'm a bit confused with the example using tolerance. Once we finalize that, this PR will be good to go.

File: examples/param_est.ipynb [hunk 17]
```diff
@@ -0,0 +1,500 @@
...
+    "    print(\"-\", key, m.parameters[key])\n",
+    "print(' Error: ', m.calc_error(times, inputs, outputs, dt=0.1))"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Notice that the error is quite high. This indicates that the parameters are not accurate\n",
```
Comment (kjjarvis) [inline examples/param_est.ipynb:151]:
```suggestion
    "Notice that the error is quite high. This indicates that the parameters are not accurate.\n",
```

File: examples/param_est.ipynb [hunk 18]
```diff
@@ -0,0 +1,500 @@
...
+    "\n",
+    "For example, if in the nth iteration of the optimizer above the best error was __2e-5__ and the cooresponding values were thrower_height=1.8, throwing_speed=40, and g=-9.8 and at the n+1th iteration the best error was __1e-5__ and the cooresponding values were thrower_height=1.85, throwing_speed=39.5, and g=-9.81, then the difference in error would be __1e-5__ and the difference in parameter values would be \n",
+    "\n",
+    "$$\\sqrt{(1.85 - 1.8)^2 + (40 - 39.5)^2 + (9 - 9.81)^2} = 0.5025932749$$\n",
+    "\n",
+    "In this case, error would meet a tol of __1e-4__, but the parameters would not, so optimization would continue. For more information, see the [scipy.optimize.minimize](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html) documentation.\n",
+    "\n",
+    "In our previous example, note that our total error was roughly __6e-10__ after the `estimate_params()` call, this corresponds to the error at the default `tol` of __1e-4__. Now, let us see what happens to the parameters when we pass a tolerance of __1e-6__."
```
Comment (kjjarvis) [inline examples/param_est.ipynb:215]:
```suggestion
    "In our previous example, note that our total error was roughly __6e-10__ after the `estimate_params()` call, using the default `tol` of __1e-4__. Now, let us see what happens to the parameters when we pass a tolerance of __1e-6__."
```

File: examples/param_est.ipynb [hunk 19]
```diff
@@ -0,0 +1,500 @@
...
+    "    print(\"-\", key, m.parameters[key])\n",
+    "print(' Error: ', m.calc_error(results.times, results.inputs, results.outputs))"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this case the error from calc_error is low, but one thing to note is to have a good estimation of the error, we should be manually measuring the Absolute Mean Error rather than using calc_error().\n",
```
[the remark this answers is not in the document]
Comment (kjjarvis) [reply examples/param_est.ipynb:331]: @aqitya This change doesn't seem to have been made. Let me know if you don't think it's a good suggestion or we should reword it in a different way.

File: tests/test_estimate_params.py (hunk 16 above)
Comment (kjjarvis) [reply tests/test_estimate_params.py:948]: @aqitya outstanding comment that needs addressing

**Review by kjjarvis — CHANGES_REQUESTED** [review]
This is looking good, @aqitya. I made just minor suggestions this time, but there are still a handful of outstanding comments from the last review. As soon as those are addressed, I think we'll be ready to approve.

File: tests/test_estimate_params.py (hunk 16 above)
Comment (aqitya) [reply tests/test_estimate_params.py:948]: Yeah, this needs to be reverted back. I was using some of the previous tests to review my understanding of tolerance. I'll revert them back.

**Review by kjjarvis — APPROVED** [review]
This looks great. Thanks, @aqitya, for being so thorough!
